### PR TITLE
Implicit conversions for put/get

### PR DESCRIPTION
**Pull request by gf712**
Implicit conversions as discussed in #5045. Would have to add all the primitive types, and probably come up with some better abstraction. This does the same as the Any::CastingRegistry, but by putting all this logic in SGObject we can add some very specific edge cases, like `std::variant`. I think this separates `Any` as a "type safe" type erasure abstraction from the parameters that use it.

### CODE REVIEW

File: tests/unit/base/SGObject_unittest.cc [hunk 1]
```diff
@@ -483,9 +483,10 @@ TEST(SGObject, tags_set_get_int)
 
 	EXPECT_THROW(obj->get<int32_t>("foo"), ShogunException);
 	obj->put(MockObject::kInt, 10);
+	EXPECT_NO_THROW(obj->put(MockObject::kInt, 10.0));
 	EXPECT_EQ(obj->get(Tag<int32_t>(MockObject::kInt)), 10);
-	EXPECT_THROW(obj->get<float64_t>(MockObject::kInt), ShogunException);
-	EXPECT_THROW(obj->get(Tag<float64_t>(MockObject::kInt)), ShogunException);
+	EXPECT_EQ(obj->get<float64_t>(MockObject::kInt), 10.0);
+	EXPECT_EQ(obj->get(Tag<float64_t>(MockObject::kInt)), 10.0);
```
Comment (gf712) [inline tests/unit/base/SGObject_unittest.cc:489]: @karlnapf so this won't throw anymore

Comment (karlnapf) [reply tests/unit/base/SGObject_unittest.cc:489]: yep.
But everything that looses information throws?
maybe a test for some examples?

Comment (gf712) [reply tests/unit/base/SGObject_unittest.cc:489]: by lose information you mean something like 1.3 becomes 1? Right now I think this function only supports checks for overflow, e.g. int -10 is converted to unsigned int.

Comment (karlnapf) [reply tests/unit/base/SGObject_unittest.cc:489]: Yes, we wouldnt want that to happen for now I think.

Comment (gf712) [reply tests/unit/base/SGObject_unittest.cc:489]: @vigsterkr would it make sense to extend safe_convert to check floating point casting to integers and see if the delta is larger than fepsilon?

Comment (karlnapf) [reply tests/unit/base/SGObject_unittest.cc:489]: I always thought that the safe_convert functions would be for overflow checks, not for loss of information when going from discrete to floating?

Comment (karlnapf) [reply tests/unit/base/SGObject_unittest.cc:489]: (on the other hand, why not :) )
